src/main/cljc/jarl/builtins/http.cljc

```diff
@@ -0,0 +1,199 @@
+
+(defn verify-url [s]
+  (if (nil? (re-find #"https?://(\S+)" s))
+    ; note: does not mimic OPA error messages yet
+    (throw (errors/builtin-ex "http.send: invalid URL: %s" s))
+    s))
+
+(defn- to-json [x]
+  (if (string? x)
+    x
+    (json/write-str x)))
+
+(def ^:private unit->nano-map
+  {"h"  3600000000000
+   "m"  60000000000
+   "s"  1000000000
+   "ms" 1000000
+   "us" 1000
+   "µs" 1000
+   "ns" 1})
+
+(defn time-unit->nanos [[_ time unit :as time-unit]]
+  (let [time (parse-double time)]
+    (if (or (nil? time) (not (contains? unit->nano-map unit)))
+      (throw (errors/builtin-ex "http.send: invalid timeout value: %s" time-unit))
+      (* time (get unit->nano-map unit)))))
+
+(defn- nanos->millis [nanos]
+  (max 1 (math/round (/ nanos 1000000))))
+
+(defn parse-timeout [x]
+  (if (number? x)
+    x
+    (if (and (string? x) (utils/numeric? x))
+      (parse-long x)
+      (if-let [time-unit-pairs (re-seq #"([\.\d]+)(\p{L}+)" x)]
+        (nanos->millis (reduce + (map time-unit->nanos time-unit-pairs)))
+        (throw (errors/builtin-ex "http.send: invalid timeout value %s" x))))))
+
+(defn create-request [request]
+  (let [req  (partial get request)
+        has? (partial contains? request)]
+    (cond-> {:method               (keyword (str/lower-case (req "method")))
+             :url                  (verify-url (req "url"))
+             :follow-redirects     false
+             :force-json-decode    false
+             :force-yaml-decode    false
+             :insecure?            false
+             :raise-error          true
+             :conn-timeout         default-req-timeout}
+            (has? "enable_redirect")          (assoc :follow-redirects (req "enable_redirect"))
+            (has? "body")                     (assoc :body (to-json (req "body")))
+            (has? "raw_body")                 (assoc :body (req "raw_body"))
+            (has? "headers")                  (assoc :headers (req "headers"))
+            (has? "timeout")                  (assoc :conn-timeout (parse-timeout (req "timeout"))) ; millis here, not OPA nanos
+            (has? "force_json_decode")        (assoc :force-json-decode (req "force_json_decode"))
+            (has? "force_yaml_decode")        (assoc :force-yaml-decode (req "force_yaml_decode"))
+            (has? "raise_error")              (assoc :raise-error (req "raise_error"))
+            (has? "tls_insecure_skip_verify") (assoc :insecure? (req "tls_insecure_skip_verify")))))
+
+(defn content-type [response]
+  (when-let [ct (get-in response [:headers "content-type"])]
+    (let [stop (or (str/index-of ct ";") (count ct))]
+      (str/lower-case (subs ct 0 stop)))))
+
+(defn response-body-decoded [request response]
+  (let [ct (content-type response)]
+    (cond
+      (:force-json-decode request) (json/read-str (:body response))
+      (:force-yaml-decode request) (yaml/read-str (:body response))
+      (= ct "application/json")    (json/read-str (:body response))
```

Comment:
A pretty common pattern is to use the `+json` suffix (or what have you) to differentiate on types, but still signal on format. E.g. `application/error+json`. Would it be too fuzzy to accommodate this, and try to parse all `application/*+json` as json and all `application/*+yaml` as yaml?

Reply:
Unless we do that in OPA — and I don't think we do? — we probably don't want that here, or WDYT?

Reply:
I'm all for improvements (they could even trickle back to OPA if good enough) :). But maybe any larger divergences from how OPA functions should be kept behind some gatekeeper flag.